### PR TITLE
Add health check endpoint

### PR DESCRIPTION
**Pull request by ttys3**
Add `/api/healthz` for health check. Which does not need auth and is cloud native friendly

By convention, most cases the path is `/healthz` or `/_healthz`, I think we need this.

https://kubernetes.io/docs/reference/using-api/health-checks/

https://www.nomadproject.io/docs/job-specification/service#header-stanza

### CODE REVIEW

**Review by zeripath — DISMISSED** [review]
This will require a new username restriction,

Is there really nothing we can use on say /api or even in .well-known?

If it has to be this name then you must add a restriction to models/user and mark this as BREAKING.

**Comment by ttys3** [comment]
@zeripath no, the path (`/healthz` or `/_healthz`) is only for convention.

do you think change it o `/api/v1/_healthz` is OK ?

**Comment by lunny** [comment]
We already have `HEAD /`

**Comment by ttys3** [comment]
`HEAD /` is not as clear as a standalone `GET /_healthz`.
and also, not all health check has builtin `HEAD` method option. although you can use a custom http client like `curl`, but this is not normal way.

adding an explicit `GET /_healthz`  will always be good and has better compatibility

**Comment by techknowlogick** [comment]
Fwiw, the gitea project uses the version api endpoint for this in our monitoring

File: routers/init.go
Comment (lafriks) [inline routers/init.go]: Shouldn't it do actual health check, let's say at least ping database?

Comment (ttys3) [reply routers/init.go]: the whole thing is for container orchestration to check the app is live.
It can be in different level. this is a simple `200` code response, the very basic level says that: hey, I'm ready and I'm not down. in most case this is enough. because if the app panic or segmentfault exited, the orchestration has a reliable way to ensure it is died.

We can also add `redis` ping and `database` ping check, but this will have a little overhead. And this, in other way,
it can be other kind of `live` or `healthz` for the db itself.  Ask the app to check seems not needed.
if redis is down, the alert should send from the redis monitor (like promethus blackbox probe alert).

and what's more, the app (in here it is gitea golang app) can have metrics for alert system to use, in that metrics, it can report the `redis` or `database` connection error. and when it exceeded specific number, the alert system will trigger an alert

so, a simple `200` (for the probe system) + ok (for human) is what I recommend.

Comment (ttys3) [reply routers/init.go]: currently, I'm using `/assets/img/logo.svg?from-health-check` as the check path.
Because 1. it does not need api authentication. simple to deploy  2. logo.svg response in text and has small body.

But, it looks very confusing. so I think we need a simple healthz endpoint like sentry, grafana, k8s etc...

Comment (ttys3) [reply routers/init.go]: [traefik](https://github.com/traefik/traefik) has `/ping`, and it just returns 200 code and an ok just like this.
do you mind if I change the path to `/api/_ping`  or `/api/ping` ?

Comment (6543) [reply routers/init.go]: I'm for `/ping` as it only return ok ...

**Comment by 6543** [comment]
I'm for `/ping` as it only return ok ... & yes it's usefully for container orchestrators ... but you could also just use some static asset served by gitea .... (but that would add unnecessary load to gitea depending on what you try to get) ...

**Comment by 6543** [comment]
& can you mention it within the docs?

**Review by 6543 — CHANGES_REQUESTED** [review]
as per tech ... use the version endpoint it does the same (return 200 & a tiny string)

It would only be a valid pull if it really do check health of the running instance (check DB, cache, storage backends, ...

File: routers/init.go
Comment (ttys3) [reply routers/init.go]: ok, I've changed it to `/api/ping`, we can not use `/ping` because this need to avoid username confliction.

**Comment by ttys3** [comment]
> version

yes, I tried the `/api/v1/version`

The version api does need authentication.  So it does not meet the needs.

returns `403` and json `{"message":"Only signed in user is allowed to call APIs."}`

I know how to build a request to pass the authentication. but this is not a `ping` api shoud look like.

when system op deploy the app, it does not and has no need to know the user account of the system.

File: routers/init.go
Comment (6543) [reply routers/init.go]: haha yes i just did now write the prefix as it was clear to me :D

**Review by 6543 — COMMENTED** [review]
ACK for code

but we need documentation

**Comment by ttys3** [comment]
@6543  some tips ? Sorry, I don't know where to add the documentation. 

Since it is not an API for client request, I think it is Ok not write swagger doc.
Maybe in the doc markdown ? But, which section ?

**Comment by 6543** [comment]
yes swagger docs would be the wrong place :D

-> https://github.com/go-gitea/gitea/tree/main/docs (just read the readme in there - it's a hugo repo)

add it to https://docs.gitea.io/en-us/install-on-kubernetes/ ?

**Comment by silverwind** [comment]
Consider implementing a response as per https://datatracker.ietf.org/doc/html/draft-inadarei-api-health-check.

**Comment by lunny** [comment]
LGTM if it's `/api/xxxx`.

File: routers/init.go
Comment (KN4CK3R) [inline routers/init.go]: The `WriteHeader` call is not needed.

Comment (ttys3) [reply routers/init.go]: OK, removed

**Comment by 6543** [comment]
docs missing :D

**Comment by ttys3** [comment]
I will come back to finish this soon. Busy work since the holiday 😂

**Comment by lunny** [comment]
The endpoint should not expose privacy information.

File: docs/content/doc/installation/on-kubernetes.en-us.md
Comment (lunny) [inline docs/content/doc/installation/on-kubernetes.en-us.md:31]: If we can give an example output, it's better.

File: routers/web/web.go
Comment (silverwind) [inline routers/web/web.go]: Why `healthz` and not just `health`?

Comment (ttys3) [reply routers/web/web.go]: Using `healthz` is just kind of  personal preference or maybe convention.

I took it because k8s itself use `healthz`, and in our daily dev, we also use `healthz`

see https://kubernetes.io/docs/reference/using-api/health-checks/
or 
https://www.nomadproject.io/docs/job-specification/service#header-stanza

maybe there's other reason, see https://github.com/jupyterhub/jupyterhub/issues/2532#issuecomment-485994741

or this https://stackoverflow.com/questions/43380939/where-does-the-convention-of-using-healthz-for-application-health-checks-come-f/43381061#43381061

> It historically comes from Google’s internal practices. They're called "z-pages".

> The reason it ends with z is to reduce collisions with actual application endpoints with the same name (like /status). 

but there's also other examples, like grafana https://grafana.com/docs/grafana/latest/http_api/other/#health-api
it is `/api/health`

Comment (ttys3) [reply routers/web/web.go]: so I think we use `healthz` here is better, means it is not a real api. (but we mount it under `/api`)

Comment (silverwind) [reply routers/web/web.go]: Okay, fine with me, I wasn't aware of this convention.

**Comment by ttys3** [comment]
Updated,   @lunny  @silverwind  PTAL

File: modules/cache/cache.go
Comment (lunny) [inline modules/cache/cache.go:47]: Function comment is required.

File: routers/web/healthcheck/check.go
Comment (lunny) [inline routers/web/healthcheck/check.go]: Comment

Comment (ttys3) [reply routers/web/healthcheck/check.go]: done

File: modules/cache/cache.go
Comment (ttys3) [reply modules/cache/cache.go:47]: done

**Comment by ttys3** [comment]
@lunny comments added. and I also removed the `releaseId` field from the final response. gitea version should not included in the response. since it maybe related to privacy.

**Comment by ttys3** [comment]
update:

removed the `gitea:installed` status report. we should not include function check in the health check endpoint.

otherreason is: 
for the first time (before we run the installation setup in the web page), when the k8s pod startup and gitea service start running,
the liveness probe will get false `fail` status report since gitea has not finishing the installation steps.

**Comment by ttys3** [comment]
now the respons is like:

```
HTTP/1.1 200 OK
Content-Type: application/json


{
  "status": "pass",
  "description": "Gitea: Git with a cup of tea",
  "checks": {
    "cache:ping": [
      {
        "status": "pass",
        "time": "2022-02-19T11:46:14Z"
      }
    ],
    "database:ping": [
      {
        "status": "pass",
        "time": "2022-02-19T11:46:14Z"
      }
    ]
  }
}
```

File: routers/web/healthcheck/check.go
Comment (lunny) [inline routers/web/healthcheck/check.go]: Sorry, I think lint should have some problem. All public variables and functions should have comments.

Comment (ttys3) [reply routers/web/healthcheck/check.go]: Yes, I think so.

But  now I realized that I should not expose those types . So, the comments can just be avoided. -_- 

pushed, PTAL

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
need @6543 's review again.

File: modules/cache/cache.go
Comment (silverwind) [inline modules/cache/cache.go:68]: Redis has a [PING](https://redis.io/commands/PING) command, but I guess we can't use it in this abstraction, right?

File: modules/cache/cache.go
Comment (6543) [reply modules/cache/cache.go:49]: -> the first idea could result in a smal refactor but would be better, the second would work as it do things as they work now

File: modules/cache/cache.go
Comment (6543) [reply modules/cache/cache.go:68]: I would prefere that too ... but this will need a pull to https://gitea.com/go-chi/cache/src/branch/master/cache.go#L24 so it can be passed down to it's implementations ... and then this "cache ping" hack can be used there if the underlying implementation do not have a ping ...

File: routers/web/web.go
Comment (6543) [inline routers/web/web.go]: why did you implement it the "single tone" way? you dont have to

**Comment by 6543** [comment]
summary:

how the ping is done ... I dont like it, but that's how it works without touching lot of code ... (we can refactor that ... but need mentioned preparations ...) - so there is my ok for that  

but, the implementation of the check function just do not need a struct ... just remove it and that's it - why do we need a single tone ...

File: routers/web/web.go
Comment (ttys3) [reply routers/web/web.go]: yes, it does not have to.
I only use it to organize the code.
but if you do not like it, I can remove it.
thus, the `cache` and `database` func will conflict with the mod, I'll have to add prefix to them.

Comment (ttys3) [reply routers/web/web.go]: done and pushed.

File: modules/cache/cache.go
Comment (silverwind) [reply modules/cache/cache.go:68]: Ok, maybe something to do later to not complicate this here further.

**Comment by silverwind** [comment]
Actually I think we should use `/api/health`. `healthz` is [meant for load-balancers](https://stackoverflow.com/questions/43380939/where-does-the-convention-of-using-healthz-for-application-health-checks-come-f), not for applications like gitea.

**Comment by ttys3** [comment]
> Actually I think we should use `/api/health`. `healthz` is [meant for reverse proxies](https://stackoverflow.com/a/43381061/808699), not for applications like gitea.
> 
> Edit: See [jupyterhub/jupyterhub#2532 (comment)](https://github.com/jupyterhub/jupyterhub/issues/2532#issuecomment-485994741) for some info. While it appears k8s does by default discover a `/healthz`, no other cloud provider does this and even the k8s api itself violates the principle by exposing a `/healthz` as it is not a reverse proxy. So I think there is no reason to follow this "standard".

I do not think `healthz` meant for reverse proxies. even the link you provided.

There's no `reverse`, nor `proxies` in the discussion.

**Comment by ttys3** [comment]
> 

I've made changes according to the second review comment.

About the first one, yes, a new `Ping` api is better. since currently I just extract the code from the exists code and named it to `Ping()`

**Comment by silverwind** [comment]
> There's no reverse, nor proxies in the discussion.

It says "The reason it ends with z is to reduce collisions with actual application endpoints with the same name (like /status)", which I interpret that "z" endpoints are for non-applications and non-z ones are for applications.

**Comment by silverwind** [comment]
Anyways, we can keep `healthz`, even thought i don't exactly like it.

File: modules/cache/cache.go
Comment (ttys3) [reply modules/cache/cache.go:68]: @6543
since  `mc.Cache`  interface is external, I can not change it and add a `Ping` func.

 Is it OK if I  add a `Ping` func and create a new `cache` interface ?
The new interface will combine `mc.Cache` and a `Ping` func.

Comment (silverwind) [reply modules/cache/cache.go:68]: Should we open an issue about it on https://gitea.com/go-chi/cache?

Comment (6543) [reply modules/cache/cache.go:68]: yes !

Comment (ttys3) [reply modules/cache/cache.go:68]: just for mark: https://gitea.com/go-chi/cache/issues/3

**Review by 6543 — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Conflict resolved. Prepare to merge.

**Comment by lafriks** [comment]
Will this endpoint be available for installation stage?

**Comment by wxiaoguang** [comment]
> > Will this endpoint be available for installation stage?
> 
> IMO it won't cause problem for the installation stage. The API may only returns an error because there is no database connection.

Update: there should be no problem. On the installation stage, only routes in `install.go` are available.

**Comment by lafriks** [comment]
I mean than this can not be used for docker as health check as container won't start when installation stage is not completed

**Comment by wxiaoguang** [comment]
> I mean than this can not be used for docker as health check as container won't start when installation stage is not completed

Sorry I can not get the point. Why this end point should be used when installation stage is not completed? Can you use a real example to help me to understand the details?

**Comment by lafriks** [comment]
If I would add to dockerfile:
```
HEALTHCHECK CMD curl --fail http://localhost:3000/api/healthz || exit 1
```
And deploy new instance (unconfigured thus only /install page accessible) than I won't be able to access that container as as it won't never become in healthy state

**Comment by wxiaoguang** [comment]
But this end point is not for docker, it's for k8s. When using it in k8s (cloud native), the instance should have been ready by helm-chart or something else.

Docker user should only check the home page, not the api healthz

**Comment by lafriks** [comment]
I understand that but for docker it need check only http status code not the content

**Comment by wxiaoguang** [comment]
> I understand that but for docker it need check only http status code not the content

Then we have 2 choices:

1. Current implement: the `/api/healthz` only works for installed stage. For dokcer users, use the `/` or `/api/v1/version` instead, since the HEALTHCHECK CMD doesn't read the response content.
2. Further improvement: make `/api/healthz` work both for installation stage and the installed stage. On the installation stage (by checking setting.InstallLock), the healthz end point should ignore all checks, just return status code 200

If I understand correctly, you suggest to do solution 2 before merge, right?

**Comment by lafriks** [comment]
Yes I was thinking about option 2 as first one would return healthy even if no redis or db was actually accessible.

It does not leak private information directly but could theoretically be abused to create load on cache and database even if instance is set as private but this is just more of a suggestion and not blocking for this PR

**Comment by wxiaoguang** [comment]
> It does not leak private information directly but could theoretically be abused to create load on cache and database even if instance is set as private

However, that's the same to other API endpoints ...... attackers can do CC attack to the API endpoints which needs authorization, the login check is much more heavy than the health check .....

**Comment by lafriks** [comment]
For private instances you can easily set fail2ban to counter that

**Comment by ttys3** [comment]
@wxiaoguang option 2 will works, but maybe a little confusing ?
should we add another `/readyz` that just return a simple `OK` ? which will works for not installed gitea instance.

the diff between other micro service is, gitea need interactive installation to work for the first time.

is there any automatic method to make this work in k8s without interactive installation? I man something like a cli to allow fast installation

**Comment by jbg** [comment]
Does Gitea actually need interactive installation the first time? I thought you could fill in the config and enable install lock and it would just work (run the migrations on first launch).

**Comment by wxiaoguang** [comment]
> @wxiaoguang option 2 will works, but maybe a little confusing ? should we add another `/readyz` that just return a simple `OK` ? which will works for not installed gitea instance.
> 
> the diff between other micro service is, gitea need interactive installation to work for the first time.
> 
> is there any automatic method to make this work in k8s without interactive installation? I mean something like a cli to allow fast installation (for example: `gitea installation`), thus, the instance will initilzation itself, so the healthz api should also work for the first time.

@lafriks

**Comment by lafriks** [comment]
It's not about if healthz endpoint being usable for k8s, for k8s there is helm chart that will make gitea usable without installation. Problem is that health endpoint is not usable for docker/docker swarm, at least from most used perspective install stage is most common scenario as we don't have easy way to skip it.

**Comment by lafriks** [comment]
I don't see much point in readyz endpoint as that would pretty much be same as `HEAD /` request. Point for healthz endpoint is that requests can be started to route to that container as it has database and cache available and container is in working condition

**Comment by markkrj** [comment]
Couldn't one `/api/healthz` endpoint be registered in installation routes that will ever return 200? That way, if Gitea isn't provisioned, Docker will think it's healthy, and if it is, installation routes will not be used anyway...

**Comment by lafriks** [comment]
> Couldn't one `/api/healthz` endpoint be registered in installation routes that will ever return 200? That way, if Gitea isn't provisioned, Docker will think it's healthy, and if it is, installation routes will not be used anyway...

Yes, that was my suggestion

**Comment by markkrj** [comment]
> > Couldn't one `/api/healthz` endpoint be registered in installation routes that will ever return 200? That way, if Gitea isn't provisioned, Docker will think it's healthy, and if it is, installation routes will not be used anyway...
> 
> Yes, that was my suggestion

@lafriks @wxiaoguang would ttys3/gitea#1 be acceptable, then?

**Comment by wxiaoguang** [comment]
@lafriks

**Comment by markkrj** [comment]
> @lafriks

Or maybe you just want `r.Get("/api/healthz", func(http.ResponseWriter, *http.Request){})` on `install` routes, to not depend on `healthcheck` package... That might be even better/simpler..

File: modules/cache/cache.go
Comment (lafriks) [inline modules/cache/cache.go]: Would be better to use `if err = Ping(); err != nil {`

**Review by lafriks — COMMENTED** [review]
Small nit but otherwise looks good

File: modules/cache/cache.go
Comment (wxiaoguang) [reply modules/cache/cache.go]: ~~No, ping doesn't work (can not tell that) if the cache if full.~~

Oh, I see ... 😂

Comment (wxiaoguang) [inline modules/cache/cache.go]:
```suggestion
		if err = Ping(); err != nil {
```

Comment (wxiaoguang) [reply modules/cache/cache.go]: I see the point, updated. `Ping` already does what it should do.

**Review by wxiaoguang — APPROVED** [review]
OK, since there is no objection, let me take the remaining work:
1. conflicts are resolved
2. remove `CheckInstall`, use `Check` with InstallLock check instead.
3. set `Description` to `setting.AppName` instead of the hard-coded string.

**Review by lafriks — APPROVED** [review]

**Pull request merged**
